File: typescript/Behavioral/22-Visitor/Element/PlumbingJob.ts
import type { IJobElement } from "./IJobElement";
import type { IJobVisitor } from "../Visitor/IJobVisitor";

export class PlumbingJob implements IJobElement {
  readonly jobId: string = Math.random().toString(36).slice(2, 10);

  constructor(
    readonly title: string,
    readonly cost: number,
    readonly pressureBar: number,
    readonly pipeMetres: number,
  ) {}

  // Double dispatch — calls back into the visitor's plumbing method
  accept(visitor: IJobVisitor): void {
    visitor.visitPlumbing(this);
  }
}

File: csharp/Behavioral/22-Visitor/Element/PlumbingJob.cs
namespace Visitor;

public class PlumbingJob : IJobElement
{
    public string JobId { get; } = Guid.NewGuid().ToString("N")[..8];

    public string Title { get; }
    public decimal Cost { get; }
    public decimal PressureBar { get; }
    public decimal PipeMetres { get; }

    public PlumbingJob(string title, decimal cost, decimal pressureBar, decimal pipeMetres)
    {
        Title       = title;
        Cost        = cost;
        PressureBar = pressureBar;
        PipeMetres  = pipeMetres;
    }

    // Double dispatch — calls back into the visitor's PlumbingJob overload
    public void Accept(IJobVisitor visitor) => visitor.Visit(this);
}
